### ISSUE TITLE
Different behaviours when cancelling a process in Windows and Linux

### ISSUE THREAD

**Issue by capitanbatata** [Gabriel439/Haskell-Turtle-Library#255]
Cancelling an asynchronous process leads to different behavior depending on the host OS. I've made [an example that shows this](https://github.com/capitanbatata/sandbox/tree/master/racing-turtles). The test program presented in this example spawns an external process that runs forever (see `SayHi.java`) and after a certain time it cancels it (by using `race`). In the test program all the functions (`internalExample`, `procExample`, `forkAsyncExample`, `inProcExample`, and `sysProcExceptionExample`) terminate the Java program. However in Windows 10 `procExample` and `forkAsyncExample` do not terminate.

**Comment by Gabriel439** [comment]
My intuition is that this is not an issue with the `turtle` library but rather an issue with either the `async` or `process` library.  For example, here's a patch that simplifies `forkAsyncExample` to not use `turtle` at all:

```diff
diff --git a/racing-turtles/src/Lib.hs b/racing-turtles/src/Lib.hs
index 69f126b..1c09dc5 100644
--- a/racing-turtles/src/Lib.hs
+++ b/racing-turtles/src/Lib.hs
@@ -41,9 +41,7 @@ inProcExample = mkExample ip
 
 forkAsyncExample :: IO ()
 forkAsyncExample = mkExample forkAsyncHighlander
-  where forkAsyncHighlander = sh $ using $ do
-          a <- fork highlander
-          Turtle.wait a
+  where forkAsyncHighlander = withAsync highlander Control.Concurrent.Async.wait
 
 sysProcExceptionExample :: IO ()
 sysProcExceptionExample = mkExample aMereMortalProc
```

The other reason that I suspect that this is not a `turtle` issue is that `turtle` doesn't have different code for Windows or Linux when it comes to subprocess management or concurrency

I can't test on Windows at the moment, but you can check to see if you get the same results with the above patch

**Comment by capitanbatata** [comment]
With the change above the `forkAsyncExample` still does not terminates on Windows, since it is still using `proc` from `turtle`:

```haskell
highlander = void (proc "java" ["-cp", "src", "SayHi"] empty)
```

Strangely, `inProcExample`, which uses `inproc`, terminates on Windows:

```haskell
inProcExample = mkExample ip
  where ip = view $ inproc "java" ["-cp", "src", "SayHi"] empty
```

I've looked at the implementation of both, but I cannot see what is the difference between `system` and `stream` that might be causing this...

```haskell
stream p s = do
    let p' = p
            { Process.std_in  = Process.CreatePipe
            , Process.std_out = Process.CreatePipe
            , Process.std_err = Process.Inherit
            }

    let open = do
            (Just hIn, Just hOut, Nothing, ph) <- liftIO (Process.createProcess p')
            IO.hSetBuffering hIn IO.LineBuffering
            return (hIn, hOut, ph)

    -- Prevent double close
    mvar <- liftIO (newMVar False)
    let close handle = do
            modifyMVar_ mvar (\finalized -> do
                unless finalized (hClose handle)
                return True )

    (hIn, hOut, ph) <- using (managed (bracket open (\(hIn, _, ph) -> close hIn >> Process.terminateProcess ph)))
    let feedIn :: (forall a. IO a -> IO a) -> IO ()
        feedIn restore = restore (outhandle hIn s) `finally` close hIn

    a <- using
        (managed (\k ->
            mask (\restore -> withAsync (feedIn restore) (restore . k))))
    inhandle hOut <|> (liftIO (waitForProcessThrows ph *> halt a) *> empty)
```

```haskell
system p s = liftIO (do
    let open = do
            (m, Nothing, Nothing, ph) <- Process.createProcess p
            case m of
                Just hIn -> IO.hSetBuffering hIn IO.LineBuffering
                _        -> return ()
            return (m, ph)

    -- Prevent double close
    mvar <- newMVar False
    let close handle = do
            modifyMVar_ mvar (\finalized -> do
                unless finalized (ignoreSIGPIPE (hClose handle))
                return True )
    let close' (Just hIn, ph) = do
            close hIn
            Process.terminateProcess ph
        close' (Nothing , ph) = do
            Process.terminateProcess ph

    let handle (Just hIn, ph) = do
            let feedIn :: (forall a. IO a -> IO a) -> IO ()
                feedIn restore =
                    restore (ignoreSIGPIPE (outhandle hIn s)) `finally` close hIn
            mask (\restore ->
                withAsync (feedIn restore) (\a ->
                    restore (Process.waitForProcess ph) `finally` halt a) )
        handle (Nothing , ph) = do
            Process.waitForProcess ph

    bracket open close' handle )
```

**Comment by Gabriel439** [comment]
So there are two separate questions:

* Why does it hang?
* Why is the behavior platform-specific?

Also, the  problem could be in one of the following four places:

* the `turtle` library
* the `process` library
* the `async` library

So I need you to conduct two experiments to narrow things down.  The result is most likely going to point to `turtle` being at fault, but doing these experiments will also help me significantly narrow down where the issue is in the `turtle` code

# Experiment 1

You can determine if the `process` library is relevant by modifying the `highlander` function to use the `process` library directly, like this:

```haskell
highlander :: IO ()
highlander = System.Process.system "java -cp src SayHi"
```

If the problem persists, then that indicates that the `process` library is the problem

# Experiment 2

You can determine if the `async` library is relevant by changing the `highlander` function to use the `Turtle.system` function directly, but like this:

```haskell
highlander :: IO ()
highlander = void (Turtle.system (System.Process.proc "java" ["-cp", "src", "SayHi"]) empty)
```

That uses a completely `async`-free code path.  If the problem disappears then the `async` library is the problem

**Comment by capitanbatata** [comment]
Hi Gabriel. Sorry for the late reply, but I only have a Windows computer at work, and I could not find free time to carry out these experiments.

Both experiments do not terminate :/ Somehow `System.waitForProcess` cannot be killed. That's why I wrote `sysProcExceptionExample`, where instead of waiting for a process I just make the thread wait forever.

I just pushed the relevant changes to [my repository](https://github.com/capitanbatata/sandbox/tree/master/racing-turtles).

I'm also going to try with the stack nightly snapshot (which uses `process` 1.6.1.0 instead of 1.4.3.0).

**Comment by Gabriel439** [comment]
So if the first experiment did not terminate then that indicates that it's a problem with the `process` library because there is no `turtle` involved when running `System.Process.system`

**Comment by capitanbatata** [comment]
Indeed! So I'm closing this now. The problem in `System.Process.system` might have been solved in https://github.com/haskell/process/pull/101, but apparently it was not merged into 1.6.1.0 since I still experience the same behavior with this version...
